### ISSUE TITLE
Media Gallery Query is slightly off.

### ISSUE THREAD

**Issue by pmathbliss** [magento/magento2#7527]
### Preconditions

1. Assign images to catalog_product_entity_media_gallery
2. Add records to catalog_product_entity_media_gallery_value for each value_id,entity_id that needs the image to the admin store. Assigning main images to configurable products and color images to the child simples of the configurable product.
3. Add records for each value_id, entity_id to catalog_product_entity_media_gallery_value_to_entity

### Steps to reproduce

1. Open the configurable product page. 
2. The configurable will show all images. Including the simples images. 

### Expected result

1. The configurable should only have its images assigned to the entity_id in catalog_product_entity_media_gallery_value and catalog_product_entity_media_gallery_value_to_entity
2. This query show the distinct images and not images assigned to simples. 
>SELECT 
  `main`.`value_id`,
  `main`.`value` AS `file`,
  `main`.`media_type`,
  `entity`.`entity_id`,
  `value`.`label`,
  `value`.`position`,
  `value`.`disabled`,
  `default_value`.`label` AS `label_default`,
  `default_value`.`position` AS `position_default`,
  `default_value`.`disabled` AS `disabled_default`,
  `value_video`.`provider` AS `video_provider`,
  `value_video`.`url` AS `video_url`,
  `value_video`.`title` AS `video_title`,
  `value_video`.`description` AS `video_description`,
  `value_video`.`metadata` AS `video_metadata`,
  `default_value_video`.`provider` AS `video_provider_default`,
  `default_value_video`.`url` AS `video_url_default`,
  `default_value_video`.`title` AS `video_title_default`,
  `default_value_video`.`description` AS `video_description_default`,
  `default_value_video`.`metadata` AS `video_metadata_default` 
FROM
  `catalog_product_entity_media_gallery` AS `main` 
  INNER JOIN `catalog_product_entity_media_gallery_value_to_entity` AS `entity` 
    ON main.value_id = entity.value_id 
  LEFT JOIN `catalog_product_entity_media_gallery_value` AS `value` 
    ON main.value_id = value.value_id 
    AND entity.entity_id = value.entity_id 
    AND value.store_id = 1 
  INNER JOIN `catalog_product_entity_media_gallery_value` AS `default_value` 
    ON main.value_id = default_value.value_id 
    AND default_value.entity_id = entity.entity_id
    AND default_value.store_id = 0 
  LEFT JOIN `catalog_product_entity_media_gallery_value_video` AS `value_video` 
    ON value.value_id = value_video.value_id 
    AND value.store_id = value_video.store_id 
  LEFT JOIN `catalog_product_entity_media_gallery_value_video` AS `default_value_video` 
    ON default_value.value_id = default_value_video.value_id 
    AND default_value.store_id = default_value_video.store_id 
WHERE (main.attribute_id = '90') 
  AND (main.disabled = 0) 
  AND (entity.entity_id = '20977') 
ORDER BY IF(
    value.position IS NULL,
    default_value.position,
    value.position
  ) ASC <

### Actual result

1. This is the actual query from magento currently.

> SELECT 
  `main`.`value_id`,
  `main`.`value` AS `file`,
  `main`.`media_type`,
  `entity`.`entity_id`,
  `value`.`label`,
  `value`.`position`,
  `value`.`disabled`,
  `default_value`.`label` AS `label_default`,
  `default_value`.`position` AS `position_default`,
  `default_value`.`disabled` AS `disabled_default`,
  `value_video`.`provider` AS `video_provider`,
  `value_video`.`url` AS `video_url`,
  `value_video`.`title` AS `video_title`,
  `value_video`.`description` AS `video_description`,
  `value_video`.`metadata` AS `video_metadata`,
  `default_value_video`.`provider` AS `video_provider_default`,
  `default_value_video`.`url` AS `video_url_default`,
  `default_value_video`.`title` AS `video_title_default`,
  `default_value_video`.`description` AS `video_description_default`,
  `default_value_video`.`metadata` AS `video_metadata_default` 
FROM
  `catalog_product_entity_media_gallery` AS `main` 
  INNER JOIN `catalog_product_entity_media_gallery_value_to_entity` AS `entity` 
    ON main.value_id = entity.value_id 
   LEFT JOIN `catalog_product_entity_media_gallery_value` AS `value` 
    ON main.value_id = value.value_id 
    AND value.store_id = 0 
  LEFT JOIN `catalog_product_entity_media_gallery_value` AS `default_value` 
    ON main.value_id = default_value.value_id 
    AND default_value.store_id = 0 
  LEFT JOIN `catalog_product_entity_media_gallery_value_video` AS `value_video` 
    ON value.value_id = value_video.value_id 
    AND value.store_id = value_video.store_id 
  LEFT JOIN `catalog_product_entity_media_gallery_value_video` AS `default_value_video` 
    ON default_value.value_id = default_value_video.value_id 
    AND default_value.store_id = default_value_video.store_id 
WHERE (main.attribute_id = '90') 
  AND (main.disabled = 0) 
  AND (entity.entity_id = '20977') 
ORDER BY IF(
    value.position IS NULL,
    default_value.position,
    value.position
  ) ASC <

The fix is to change the following method. Magento\Catalog\Model\ResourceModel\Product\Gallery::createBatchBaseSelect
Changes in bold.(**)
> public function createBatchBaseSelect($storeId, $attributeId)
    {
        $linkField = $this->metadata->getLinkField();

        $positionCheckSql = $this->getConnection()->getCheckSql(
            'value.position IS NULL',
            'default_value.position',
            'value.position'
        );

        $mainTableAlias = $this->getMainTableAlias();

        $select = $this->getConnection()->select()->from(
            [$mainTableAlias => $this->getMainTable()],
            [
                'value_id',
                'file' => 'value',
                'media_type'
            ]
        )->joinInner(
            ['entity' => $this->getTable(self::GALLERY_VALUE_TO_ENTITY_TABLE)],
            $mainTableAlias . '.value_id = entity.value_id',
            [$linkField]
        )->joinLeft(
            ['value' => $this->getTable(self::GALLERY_VALUE_TABLE)],
            implode(
                ' AND ',
                [
                    $mainTableAlias . '.value_id = value.value_id',
					**'entity.entity_id = value.entity_id',**
                    $this->getConnection()->quoteInto('value.store_id = ?', (int)$storeId),
                ]
            ),
            ['label', 'position', 'disabled']
        )->**joinInner**(
            ['default_value' => $this->getTable(self::GALLERY_VALUE_TABLE)],
            implode(
                ' AND ',
                [
                    $mainTableAlias . '.value_id = default_value.value_id',
					**'entity.entity_id = default_value.entity_id',**
                    $this->getConnection()->quoteInto('default_value.store_id = ?', Store::DEFAULT_STORE_ID),
                ]
            ),
            ['label_default' => 'label', 'position_default' => 'position', 'disabled_default' => 'disabled']
        )->where(
            $mainTableAlias . '.attribute_id = ?',
            $attributeId
        )->where(
            $mainTableAlias . '.disabled = 0'
        )->order(
            $positionCheckSql . ' ' . \Magento\Framework\DB\Select::SQL_ASC
        );

        return $select;
    }<

**Comment by veloraven** [comment]
@pmathbliss thank you for your feedback.
Please identify which version of Magento you are running

**Comment by pmathbliss** [comment]
2.1.2 community

**Comment by nbjohan** [comment]
This issue also occurs on 2.1.1 community

**Comment by leonboot** [comment]
I may be experiencing a related issue. Running the following code on a product catalog of 4 products, of which 2 products have an image and 2 products do not have an image, results in an empty set, while negating the query by using `['notnull' => true]` as a filter results in a set with both products that have an image.

```php
$objectManager = \Magento\Framework\App\ObjectManager::getInstance();

/** @var \Magento\Catalog\Model\ResourceModel\Product\CollectionFactory $productCollection */
$productCollection = $objectManager->create('Magento\Catalog\Model\ResourceModel\Product\CollectionFactory');

$collection = $productCollection->create()
    ->addAttributeToFilter('image', ['null' => true])
    ->addAttributeToSelect('*')
    ->load();
```

When I add `->getSelectSql()` to the `create()` method, change the first `INNER JOIN` from the returned query to a `LEFT JOIN` and run the query on the database, I'm seeing the expected result: the 2 products with no image assigned to them.

**Comment by nikoelgatito** [comment]
**In our case we ran into a different issue that would also be solved by applying the exact same modification @pmathbliss suggested.**

While building a custom product import script, as we deal with multiple stores using the same image for several product variations, we did not want to use the default Magento approach which consist of creating duplicated values in the `catalog_product_entity_media_gallery` table . Our opinion is that the table `catalog_product_entity_media_gallery` is solely responsible for storing a link to a unique physical file and therefore it should contain only unique values in the `value` field. 

This saves an enormous amount of rows that can't be neglected in catalogs where there are hundreds or thousands of product variations sharing the same images.

For this reason we built our import script making sure that it is importing only unique values while reusing the `value_id` of the existing records where applicable.

We encountered a wall once the import was completed where all backend product pages where the product in question was sharing images with other product were loading indefinitely and directly crashing our MYSQL database. We had to manually restart the MYSQL server each time to restore proper database functioning.

After reviewing the logs and digging into the code while testing multiple working alternatives we found that there is only one effective way to edit the actual Magento code to get it working while ensuring that everything else, every dependability as the videos for example would still work. This is to make sure that both LEFT JOIN with the `catalog_product_entity_media_gallery_value` are conducted ON matching `entity_id`, not solely ON `value_id` and `store_id`.

Once everything was tested and completely working I came here to create a bug report and stumbled on this already existing bug report while looking for other ones referring to the `Magento\Catalog\Model\ResourceModel\Product\Gallery` class. It seems that after all, this bug is affecting different functionalities and applying the suggested fix would ensure proper field relations.

## To confirm, the modifications to be applied are the following:

> magento/module-catalog/Model/ResourceModel/Product/Gallery.php::createBatchBaseSelect()

```
Line 180 | + 'entity.entity_id = value.entity_id',
    ...
Line 190 | + 'entity.entity_id = default_value.entity_id',
```

# Additional Recommendation
I would still like to recommend Magento to change the way it is adding values to the `catalog_product_entity_media_gallery` table in a way that it treats the `value` field as unique. This would save several unnecessary rows that are simply useless duplicates.

______________________________________________________________________

Magento Version 2.1.6

**Comment by pmathbliss** [comment]
yes, exactly the way I see it. The schema is set up for unique images
applied to many products but they dropped it somewhere along the
development.

**Comment by magento-engcom-team** [comment]
@pmathbliss, thank you for your report.
The issue is already fixed in 2.2.0
